org-code-javabuilder/lib/src/main/java/org/code/javabuilder/CodeExecutionManager.java

```diff
@@ -0,0 +1,182 @@
+
+  static class CodeBuilderRunnableFactory {
+    public CodeBuilderRunnable createCodeBuilderRunnable(
+        ProjectFileLoader fileLoader,
+        OutputAdapter outputAdapter,
+        File tempFolder,
+        ExecutionType executionType,
+        List<String> compileList) {
+      return new CodeBuilderRunnable(
+          fileLoader, outputAdapter, tempFolder, executionType, compileList);
+    }
+  }
+
+  public CodeExecutionManager(
+      ProjectFileLoader fileLoader,
+      InputHandler inputHandler,
+      OutputAdapter outputAdapter,
+      ExecutionType executionType,
+      List<String> compileList,
+      JavabuilderFileManager fileManager,
+      LifecycleNotifier lifecycleNotifier) {
+    this(
+        fileLoader,
+        inputHandler,
+        outputAdapter,
+        executionType,
+        compileList,
+        fileManager,
+        lifecycleNotifier,
+        new CodeBuilderRunnableFactory());
+  }
+
+  public CodeExecutionManager(
+      ProjectFileLoader fileLoader,
+      InputHandler inputHandler,
+      OutputAdapter outputAdapter,
+      ExecutionType executionType,
+      List<String> compileList,
+      JavabuilderFileManager fileManager,
+      LifecycleNotifier lifecycleNotifier,
+      CodeBuilderRunnableFactory codeBuilderRunnableFactory) {
+    this.fileLoader = fileLoader;
+    this.inputHandler = inputHandler;
+    this.outputAdapter = outputAdapter;
+    this.executionType = executionType;
+    this.compileList = compileList;
+    this.fileManager = fileManager;
+    this.lifecycleNotifier = lifecycleNotifier;
+    this.codeBuilderRunnableFactory = codeBuilderRunnableFactory;
+    this.executionInProgress = false;
+  }
+
+  /** Executes code synchronously, ensuring that pre- and post-execution steps are performed. */
+  public void execute() {
+    try {
+      this.onPreExecute();
+      try {
+        // Run code builder
+        this.executionInProgress = true;
+        final CodeBuilderRunnable runnable =
+            this.codeBuilderRunnableFactory.createCodeBuilderRunnable(
+                this.fileLoader,
+                this.outputAdapter,
+                this.tempFolder,
+                this.executionType,
+                this.compileList);
+        runnable.run();
+      } catch (Throwable e) {
+        // Catch any throwable here to ensure that onPostExecute() is always called
+        LoggerUtils.logException(e);
+      }
+      this.onPostExecute();
+    } catch (InternalServerError e) {
+      LoggerUtils.logError(e);
+    }
+  }
+
+  /**
+   * Request the code execution to cleanup early (e.g. in the case of an impending timeout). Code
+   * execution may still run, but this allows for post execution steps to still occur before the
+   * environment is shut down.
+   */
+  public void requestEarlyExit() {
+    if (!this.executionInProgress) {
+      Logger.getLogger(MAIN_LOGGER)
+          .warning("Early exit requested while execution not in progress.");
```

Comment:
Do we need both this log and the one on line 158?

Reply:
I think so - we log here in case a consumer calls `requestEarlyExit()` before calling `execute()`, and log in `onPostExecute()` if we have requested an early exit while code was running and try to call `onPostExecute()` twice (once in `requestEarlyExit()` and again after `runnable.run()` has completed - we don't want the second call to execute)